File: 36. Valid Sudoku.py
# Time and space complexity = O(n^2)
from collections import defaultdict


class Solution:
       def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = defaultdict(set) # It will create a dict with key as indexes from 0-9 and add the values as a set for respective index
        rows = defaultdict(set) # Same for Queue (key:set()) eg. {0: {'7', '3', '5'}})
        squares = defaultdict(set) # Same dictionary for the squares

        for r in range(9):      # Iterating over every element in the board
            for c in range(9):
                if board[r][c] == '.':  # If the value is . continue
                    continue
                if (board[r][c] in rows[r] or # We'll check if the current value exist in the respective row 
                    board[r][c] in cols[c] or # We'll check if the current value exist in the respective Col 
                    board[r][c] in squares[(r//3,c//3)]): # We'll check if the current value exist in the respective matrix 
                    return False    # If yes, not a valid sudoku
                cols[c].add(board[r][c])    # Adding the values to dictionaries
                rows[r].add(board[r][c])
                squares[(r//3,c//3)].add(board[r][c])
        return True
        

File: 36. Valid Sudoku.cpp
/*
Time: O(rows*cols)
Space: O(rows*cols)

For every row, column and block, maintain a separate hashset
And for every element traversed, find its row, column and block and check if the corresponding hashset contains the value already
If yes, return false
If the whole traversal is done and breach is not found, return true
*/

class Solution {
public:
    //hepler functiont to get block no. given row and column
    int get_block(int r, int c){
        return r/3 * 3 + c/3;
    }

    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_set<int> rows_set[9], cols_set[9], blocks_set[9];

        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if(board[i][j] == '.')  continue;
                int value = board[i][j];

                int row = i, col = j, block = get_block(i,j);           

                if(rows_set[row].find(value) != rows_set[row].end()
                || cols_set[col].find(value) != cols_set[col].end()
                || blocks_set[block].find(value) != blocks_set[block].end()){
                    //cout<<i<<" "<<j<<endl;
                    return false;
                }

                rows_set[row].insert(value);
                cols_set[col].insert(value);
                blocks_set[block].insert(value);

            }
        }

        return true;
    }
};
